spec/jwt_spec.rb

```diff
@@ -353,6 +353,68 @@
   end
 
   context 'Verify' do
+    context 'find_key' do
```

Comment:
Would it make sense to have a simple test case also where the keys are given via the second argument to the `JWT.decode` method?

For readability it could make sense be more explicit in what keys are used in the tests. Not maybe relying on the `iss` headers?

Reply:
> Would it make sense to have a simple test case also where the keys are given via the second argument to the JWT.decode method?

This is already covered, so I didn't want to add redundant tests and use this only for the block versions.

Reply:
> For readability it could make sense be more explicit in what keys are used in the tests. Not maybe relying on the iss headers?

Sorry, I do not understand this... The key is from the original `data` hash. The use-case I'm trying to add tests for is - each ISS can have it's own secret/key. So because I can't pass the secret upfront, because I don't know the issuer, I rely on the `find_key` block to get the issuer from the payload and use the issuer to fetch the secret from my secrets hash. I have added a direct test case as well which does something similar but without the whole issuer thing. The idea remains the same, payload contains information for me, which helps me retrieve the secret/key to verify the JWT. So I can use `find_key` for such use-cases. Let me know if this makes sense.

Reply:
Sorry for being unclear, I'll try to rephrase with an example:)

What i tired to say is that the tests are very specific to maybe your use-case, but the functionality is very generic so it can be a little confusing to understand what the tests are really testing.

I think just a set of simple tests testing the array handling would be enough:
``` 
context 'when key given as an array with multiple possible keys' do
      let(:payload) { { 'data' => 'data'} }
      let(:token)   { JWT.encode(payload, secret, 'HS256') }
      let(:secret)  { 'hmac_secret' }
  
      it 'should be able to verify signature when block returns multiple keys' do
        decoded_token = JWT.decode(token, nil, true, { algorithm: 'HS256' }) { |_headers, _payload| ['not_the_secret', secret] }
        expect(decoded_token.first).to eq(payload)
      end

      it 'should be able to verify signature when multiple keys given as a parameter' do
        decoded_token = JWT.decode(token, ['not_the_secret', secret] , true, { algorithm: 'HS256' })
        expect(decoded_token.first).to eq(payload)
      end

      it 'should fail if only invalid keys are given' do
        expect do
          JWT.decode(token, ['not_the_secret', 'not_the_secret_2'] , true, { algorithm: 'HS256' })
        end.to raise_error(JWT::VerificationError, 'Signature verification failed')
      end
    end
```

Reply:
Got it, have add your suggestions! :-)